README.md
# Implementation Services - Consul Autopilot Health Check Agent
This repository contains an application designed to execute health checks against a Hashicorp Consul Enterprise cluster. The goal of these health checks is to ensure an Autopilot-enabled cluster utilizing UpgradeVersionTag for automatic blue/green upgrades has successfully transferred voter status and cluster leadership, then output this information for observation using an outside (e.g. cloud platform) health checking service.

The impetus for developing this service lies with deploying [Consul Autopilot Upgrade Migrations](https://www.consul.io/docs/enterprise/upgrades) in environments where a signaling mechanism such as EC2 Instance Lifecycle Hooks or Azure VM Extensions does not exist. With the addition of this agent, we can observe and ensure a Consul cluster has fully transitioned voter and leader status to a set of replacement instances, before destroying the outgoing node set.

## Community Support
Support for this tool is provided by the HashiCorp Implementation Services team on a best-effort basis, and it comes with no guarantee or service level agreement. Bugs and feature requests should be entered as GitHub issues. If an existing issue matches your requirements, please add a :+1: reaction to bump it up our priority list.

Pull requests are welcome! We encourage you to fork this repository and send back a PR with any improvements.

## Configuration
A systemd unit file is included in this repository, which will launch the agent and ensure it is restarted in the event of failure. The unit file alone, however, is not enough to launch the application. It is necessary to provide a [drop-in unit](https://coreos.com/os/docs/latest/using-systemd-drop-in-units.html) to configure environment variables the application relies on to communicate with the Consul cluster.

For example, if the loaded systemd unit is named `consul-health.service`, a drop-in may look like this:

**/etc/systemd/system/consul-health.service.d/10-environment.conf:**
```
[Service]
Environment=CONSUL_HTTP_ADDR=unix:///run/consul_kv/consul_kv_http.sock
Environment=CONSUL_CLUSTER_SIZE=5
Environment=CONSUL_HEALTH_PORT=8080
```

The supported environment variables are as follows:

| Environment Variable | Default               | Description                                                                                                                            |
|----------------------|-----------------------|----------------------------------------------------------------------------------------------------------------------------------------|
| CONSUL_HTTP_ADDR     | http://127.0.0.1:8500 | Address at which Consul's HTTP listener may be reached. Prefix with `http://` for TCP transport, or `unix://` for UNIX Domain Sockets. |
| CONSUL_HTTP_TOKEN    | None                  | ACL Token to use when connecting to Consul                                                                                             |
| CONSUL_CLUSTER_SIZE  | None                  | **REQUIRED** Number of instances that should gain voter status before a cluster is considered healthy                                  |
| CONSUL_HEALTH_PORT   | 8080                  | TCP port for the health check's HTTP server to listen on                                                                               |

## Building
The latest binaries are available from the [GitHub Releases page](https://github.com/hashicorp/is-consul-health-agent/releases/latest). To build manually, run the following on a computer with go1.13+ installed:

```
git clone git@github.com:hashicorp/is-consul-health-agent.git
cd is-consul-health-agent
go build
```


CHANGELOG.md
# v0.0.2
* Add a mutex lock around the initial BoostrapHealthCheck. While this forces all requests for the initial health check to run sequentially, it ensures we can't inadvertently revert state after a successful transition to NodeHealthCheck.
* Minor refactor around environment variable lookups.

# v0.0.1
* Initial release

handlers.go
package main

import (
	"net/http"

	"github.com/hashicorp/is-consul-health-agent/health"
	log "github.com/sirupsen/logrus"
)

var isBootstrapped bool = false

func healthCheckHandler(w http.ResponseWriter, r *http.Request) {
	if !isBootstrapped {
		check := health.BootstrapHealthCheck{
			Client:      consulClient,
			ClusterSize: clusterSize,
		}
		if check.IsHealthy(r.Context()) {
			// Change global state
			// Once this has succeeded, we switch to a standard health check
			log.Info("Cluster bootstrapping succeeded! Switching to standard health check.")
			isBootstrapped = true
			w.WriteHeader(http.StatusOK)
			return
		}
		w.WriteHeader(http.StatusServiceUnavailable)
		return
	}

	check := health.NodeHealthCheck{
		Client: consulClient,
	}
	if check.IsHealthy() {
		w.WriteHeader(http.StatusOK)
		return
	}
	w.WriteHeader(http.StatusServiceUnavailable)
}


health/nodeCheck.go
package health

import (
	"github.com/hashicorp/consul/api"
	log "github.com/sirupsen/logrus"
)

// NodeHealthCheck performs a simple cluster/node sign of life test.
// /v1/status/leader is queried. In the event the node is unable to respond,
// the node is live but there is no Raft leader, or the API otherwise returns
// an error, we evaluate the node as Unehalthy.
type NodeHealthCheck struct {
	Client *api.Client
}

// IsHealthy implements the basic boolean test logic described above
func (hc *NodeHealthCheck) IsHealthy() bool {
	leader, err := hc.Client.Status().Leader()
	if err != nil {
		log.Error("Cluster status returned error: ", err)
		return false
	}
	// If we evaluate this health check on a cluster that has not
	// yet reached bootstrap_expect, this endpoint returns 200.
	// To prevent false positive, we check for empty string as well.
	if leader == "" {
		log.Error("Cluster reported no Raft leader")
		return false
	}
	return true
}


health/bootstrapCheck.go
package health

import (
	"context"
	"errors"
	"fmt"
	"strings"

	"github.com/hashicorp/consul/api"
	log "github.com/sirupsen/logrus"
)

// BootstrapHealthCheck contains the resources and methods necessary to
// confirm an autopilot-enabled Consul cluster has reached a healthy
// state and assumed cluster leadership for a given Upgrade Version Tag
type BootstrapHealthCheck struct {
	Client      *api.Client
	ClusterSize int
}

type consulServer struct {
	health api.ServerHealth
	node   *api.Node
}

// IsHealthy analyzes a Consul cluster to determine several things:
// 1. Is the cluster reporting a healthy, internally consistent state?
// 2. Does the cluster have the expected number of nodes?
// 3. Are all nodes in the current Upgrade Version Tag group voting?
// 4. Is the cluster leader a member of the current Upgrade Version Tag group?
// 5. Have all remaining servers in the cluster released voter status?
func (hc *BootstrapHealthCheck) IsHealthy(ctx context.Context) bool {
	clusterVersion, err := hc.getLocalClusterVersion()
	if err != nil {
		log.Error("Unable to determine value of Autopilot UpgradeVersionTag: ", err)
		return false
	}

	ready, err := hc.isClusterReady(ctx, clusterVersion)
	if err != nil {
		log.Error(err)
		return false
	}

	return ready
}

func (hc *BootstrapHealthCheck) getLocalClusterVersion() (string, error) {
	agent := hc.Client.Agent()
	agentStatus, err := agent.Self()
	if err != nil {
		return "", fmt.Errorf("Unable to retrieve Agent Self status: %w", err)
	}

	if version, ok := agentStatus["Meta"]["consul_cluster_version"]; ok {
		return version.(string), nil
	}

	return "", errors.New("Autopilot Upgade Version Tag missing from node configuration")
}

func (hc *BootstrapHealthCheck) isClusterReady(ctx context.Context, clusterVersion string) (bool, error) {
	//////////////////////////////////////
	// Build a list of Consul servers
	//////////////////////////////////////
	var (
		oldServers []consulServer
		newServers []consulServer
	)

	autopilotOpt := &api.QueryOptions{}
	autopilot, err := hc.Client.Operator().AutopilotServerHealth(autopilotOpt.WithContext(ctx))
	if err != nil {
		if strings.Contains(err.Error(), "429") {
			log.Info("Autopilot reports cluster is not yet in a healthy state")
			return false, nil
		}

		if strings.Contains(err.Error(), "No cluster leader") {
			log.Info("Cluster has no leader")
			return false, nil
		}

		return false, fmt.Errorf("Error querying autopilot health: %w", err)
	}

	if !autopilot.Healthy {
		// This is caught above in the 429 error
		log.Info("Autopilot reports cluster is not yet in a healthy state")
		return false, nil
	}

	filters := make([]string, len(autopilot.Servers))
	serversByID := make(map[string]api.ServerHealth)

	for i, server := range autopilot.Servers {
		filters[i] = fmt.Sprintf(`ID == "%s"`, server.ID)
		serversByID[server.ID] = server
	}

	opt := &api.QueryOptions{
		Filter: strings.Join(filters, " or "),
	}
	nodes, _, err := hc.Client.Catalog().Nodes(opt.WithContext(ctx))
	if err != nil {
		return false, fmt.Errorf("Error querying Consul Catalog Nodes: %w", err)
	}

	if len(nodes) != len(autopilot.Servers) {
		log.Info("Autopilot and Catalog report a different number of nodes. Waiting for consistency...")
		return false, nil
	}

	for _, node := range nodes {
		health, ok := serversByID[node.ID]
		if !ok {
			return false, fmt.Errorf("Node discovered in Catalog, but not matched by Autopilot: %s", node.ID)
		}

		s := consulServer{
			health: health,
			node:   node,
		}

		if node.Meta["consul_cluster_version"] == clusterVersion {
			newServers = append(newServers, s)
		} else {
			oldServers = append(oldServers, s)
		}
	}

	//////////////////////////////////////
	// Fail if we have not reached the
	// expected node count
	//////////////////////////////////////
	if len(newServers) < hc.ClusterSize {
		log.Infof("Number of new servers does not yet match expected cluster size. Total Nodes: %d Expected Nodes: %d", len(newServers), hc.ClusterSize)
		return false, nil
	}

	//////////////////////////////////////
	// Fail if autopilot has not fully
	// transferred voter status
	//////////////////////////////////////
	var foundLeader bool
	for _, server := range newServers {
		if !server.health.Voter {
			log.Infof("Replacement node is not yet voting: %s", server.health.ID)
			return false, nil
		}

		if server.health.Leader {
			foundLeader = true
		}
	}

	//////////////////////////////////////
	// Fail if we didn't see a leader in
	// the newservers set
	//////////////////////////////////////
	if !foundLeader {
		log.Info("Cluster leadership has not yet transferred to new node set")
		return false, nil
	}

	//////////////////////////////////////
	// Fail if old servers still hold
	// voter status
	//////////////////////////////////////
	for _, server := range oldServers {
		if server.health.Voter {
			log.Info("An outgoing node still holds voter status. Waiting for all old nodes to step down...")
			return false, nil
		}
	}

	return true, nil
}


main.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"strconv"
	"sync"

	"github.com/hashicorp/consul/api"
	"github.com/hashicorp/is-consul-health-agent/health"
	log "github.com/sirupsen/logrus"
)

var (
	consulClient   *api.Client
	clusterSize    int
	isBootstrapped bool = false
	mu             sync.Mutex
)

func healthCheckHandler(w http.ResponseWriter, r *http.Request) {
	if !isBootstrapped {
		mu.Lock() // We only mutate state in the initial health check, it's a one-way street
		defer mu.Unlock()

		// Check again once the lock has been acquired. If a concurrent check
		// has validated cluster state, we can short-circuit.
		if isBootstrapped {
			w.WriteHeader(http.StatusOK)
			return
		}

		check := health.BootstrapHealthCheck{
			Client:      consulClient,
			ClusterSize: clusterSize,
		}
		if check.IsHealthy(r.Context()) {
			// Change global state
			// Once this has succeeded, we switch to a standard health check
			log.Info("Cluster bootstrapping succeeded! Switching to standard health check.")
			isBootstrapped = true
			w.WriteHeader(http.StatusOK)
			return
		}
		w.WriteHeader(http.StatusServiceUnavailable)
		return
	}

	check := health.NodeHealthCheck{
		Client: consulClient,
	}
	if check.IsHealthy() {
		w.WriteHeader(http.StatusOK)
		return
	}
	w.WriteHeader(http.StatusServiceUnavailable)
}

func main() {
	var (
		err        error
		addr       string
		token      string
		listenAddr string
	)

	const (
		envClusterSize  = "CONSUL_CLUSTER_SIZE"
		envClusterAddr  = "CONSUL_HTTP_ADDR"
		envClusterToken = "CONSUL_HTTP_TOKEN"
		envListenerPort = "CONSUL_HEALTH_PORT"
	)

	if rawAddr, ok := os.LookupEnv(envClusterAddr); ok {
		addr = rawAddr
	} else {
		addr = "http://127.0.0.1:8500"
		log.Warnf("%s was not set, defaulting to %s", envClusterAddr, addr)
	}

	if rawToken, ok := os.LookupEnv(envClusterToken); ok {
		token = rawToken
	} else {
		log.Warnf("%s was not set, defaulting to no token.", envClusterToken)
	}

	if rawAddr, ok := os.LookupEnv(envListenerPort); ok {
		listenAddr = rawAddr
	} else {
		log.Warnf("%s was not set, defaulting to 8080", envListenerPort)
		listenAddr = "8080"
	}
	listenAddr = fmt.Sprintf(":%s", listenAddr)

	if raw, ok := os.LookupEnv(envClusterSize); ok {
		i, err := strconv.ParseInt(raw, 10, 32)
		if err != nil {
			log.Fatal("set cluster size: ", err)
		}
		clusterSize = int(i)
	} else {
		log.Fatalf("%s was not set. Exiting.", envClusterSize)
	}

	consulClient, err = api.NewClient(&api.Config{
		Address: addr,
		Token:   token,
	})
	if err != nil {
		log.Fatal("initialize Consul client: ", err)
	}

	log.Info("Starting up health check listener")

	http.HandleFunc("/health", healthCheckHandler)
	err = http.ListenAndServe(listenAddr, nil)
	if err != nil {
		log.Fatal("HTTP server exited: ", err)
	}
}
